src/main.rs

```diff
@@ -145,11 +199,32 @@ impl State {
     ///    * `NextStep::Win` if the user won!
     ///    * `NextStep::Continue` if the move worked, but the user didn't win
     ///    * `HanoiError::EmptyFrom` if the `mov.from` was empty
-    ///    * `HanoiError::UnstableStack` if the move tried to put a larger disk on a smaller one
+    ///    * `HanoiError::UnstableStack` if the move tried to put a larger
+    ///       disk on a smaller one
     ///
     /// No change is made to `self` if an error occurs.
     fn do_move(&mut self, mov: Move) -> Result<NextStep, HanoiError> {
-        unimplemented!()
+        let to = mov.to;
+        let from = mov.from;
+
+        // try to get the top of `from`, returning an error if it's empty
+        let from_top = try!(self.peek_disk(from)
+                            .ok_or(HanoiError::EmptyFrom(from)));
+
+        // try to push the disk onto `to`; if it succeeds, also pop
+        // that disk off of `from`
+        try!(self.push_disk(to, from_top)
+             .and_then(|()| {
+                 // unwrap is okay since we know we have something to pop
+                 self.pop_disk(from).unwrap();
+                 Ok(())
+             }));
```

Comment:
Doing the pop after the `try!(push_disk)` should be equivalent. Good to see you're comfortable with and_then though!